src/server/scripts/Draenor/Auchindoun/auchindoun.cpp

```diff
@@ -0,0 +1,397 @@
+                Talk(SAY_WELCOME_2);
+
+                task.Schedule(6s, [this](TaskContext /*task*/)
+                {
+                    me->GetMotionMaster()->MovePath(PATH_BARRIER, false);
+                });
+            });
+        }
+        else if (action == ACTION_TUULANI_BREAK_BARRIER)
+        {
+            if (!isAtBarrier || actionStarted)
+                return;
+
+            actionStarted = true;
+
+            Talk(SAY_BREAK_BARRIER_1);
+
+            _scheduler.Schedule(3s, [this](TaskContext task)
+            {
+                DoCast(SPELL_TUULANI_UNLOCK_VISUAL);
+
+                task.Schedule(4s, [this](TaskContext task)
+                {
+                    if (GameObject* holyBarrier = me->GetInstanceScript()->GetGameObject(DATA_HOLY_BARRIER))
+                        holyBarrier->SetGoState(GO_STATE_ACTIVE);
+
+                    task.Schedule(2s + 300ms, [this](TaskContext task)
+                    {
+                        Talk(SAY_BREAK_BARRIER_2);
+
+                        task.Schedule(1s + 300ms, [this](TaskContext /*task*/)
+                        {
+                            me->GetMotionMaster()->MovePath(PATH_MOVE_TO_BOSS, false);
+                        });
+                    });
+                });
+            });
+        }
+    }
+
+    void WaypointReached(uint32 waypointId, uint32 pathId) override
+    {
+        if (pathId == PATH_BARRIER)
+        {
+            switch (waypointId)
+            {
+                case POINT_TALK:
+                {
+                    _scheduler.Schedule(500ms, [this](TaskContext /*task*/)
+                    {
+                        Talk(SAY_WELCOME_3);
+                    });
+                    break;
+                }
+                case POINT_BARRIER:
+                {
+                    isAtBarrier = true;
+                    break;
+                }
+                default:
+                    break;
+            }
+        }
+        else if (pathId == PATH_MOVE_TO_BOSS)
+        {
+            if (waypointId == POINT_SOULS)
+            {
+                _scheduler.Schedule(500ms, [this](TaskContext /*task*/)
+                {
+                    me->SetFacingTo(1.727875f);
+                    Talk(SAY_HEROES_SOULS);
+                });
+            }
+        }
+    }
+
+    void UpdateAI(uint32 diff) override
+    {
+        _scheduler.Update(diff);
+    }
+
+private:
+    TaskScheduler _scheduler;
+    bool isAtBarrier = false;
+    bool actionStarted = false;
+};
+
+// 155647 - NPC Reaction
+struct at_npc_reaction : AreaTriggerAI
+{
+    at_npc_reaction(AreaTrigger* areatrigger) : AreaTriggerAI(areatrigger) { }
+
+    void OnUnitEnter(Unit* unit) override
+    {
+        if (unit->IsPlayer() || unit->IsPet() || unit->GetEmoteState() == EMOTE_STATE_READY1H_ALLOW_MOVEMENT || unit->GetEntry() == NPC_SOULBINDER_TUULANI)
+            return;
+
+        _scheduler.Schedule(1500ms, [this, unit](TaskContext task)
```

Comment:
pass guid of the unit, e.g. `[this, unitGUID = unit->GetGUID()]` and use `ObjectAccessor::GetCreature(*at, unitGUID)` to get pointer